### PR TITLE
Implemented a session storage. Also added full coverage by unit tests.

### PR DESCRIPTION
**Pull request by waltertamboer**
Please review @psren @pachico

### CODE REVIEW

**Comment by pachico** [comment]
You are not actually checking if the session has started or not.
It might be in old versions but you weren't supposed to be able to access global $_SESSION if you didn't.
Probably an exception should be thrown if you didn't, or start it yourself.

**Comment by waltertamboer** [comment]
Good point, I'll add a check to see if the sesison has started. If not I'll throw an exception. I guess starting the session is application logic?

**Comment by pachico** [comment]
I'd say it's app logic, but you could delegate it from the app to your session implementation.
In other words a:

``` php
public function __construct($namespace, $start_session = false)
// or something like it
```

File: src/Storage/Session.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    private $namespace;
+
+    /**
+     * Initializes a new instance of this class.
+     *
+     * @param string $namespace The namespace of the session.
+     */
+    public function __construct($namespace)
+    {
+        if (!$namespace) {
```
Comment (HellPat) [inline src/Storage/Session.php]: `0.3` is no valid namespace, too. check for `is_string` and leave a comment that we can typehint that if php5.x is abandoned in some years

File: src/Storage/Session.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            throw new InvalidArgumentException('The namespace is invalid.');
+        }
+
+        $this->namespace = $namespace;
+    }
+
+    /**
+     * Gets the namespace that is used for this stroage.
+     */
+    public function getNamespace()
```
Comment (HellPat) [inline src/Storage/Session.php]: should we make this private or at least protected. It's not needed and not part of the interface.
(if we decide different later it's easy to make it public. Removing from public API is hard)

File: src/Storage/Session.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    /**
+     * @inheritDoc
+     */
+    public function has($identifier)
+    {
+        if (empty($_SESSION[$this->getNamespace()])) {
+            return false;
+        }
+
+        return !empty($_SESSION[$this->getNamespace()][$identifier]);
```
Comment (HellPat) [inline src/Storage/Session.php]: use isset in this method. Edge cases could be (string) "0". But I'm not sure if it is allowed in the session

File: src/Storage/Session.php [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            return null;
+        }
+
+        return $_SESSION[$this->getNamespace()][$identifier];
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function set($identifier, $participation)
```
Comment (HellPat) [inline src/Storage/Session.php:73]: we should rename participation from all comments/variables in interface and implementation
(it's general purpose now and not limited to participations)

File: src/Storage/Session.php [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        $_SESSION[$this->getNamespace()][$identifier] = $participation;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function all()
+    {
+        if (empty($_SESSION[$this->getNamespace()])) {
```
Comment (HellPat) [inline src/Storage/Session.php]: isset?

File: src/Storage/Session.php (hunk 2 above)
Comment (waltertamboer) [inline src/Storage/Session.php]: Fair enough, I'll make it private.

File: src/Storage/Session.php (hunk 4 above)
Comment (waltertamboer) [inline src/Storage/Session.php:73]: What do you suggest?

Comment (HellPat) [inline src/Storage/Session.php:73]: $value? maybe $identifier = $key

Comment (waltertamboer) [inline src/Storage/Session.php:73]: Let's implement this in a separate branch so we can implement it for all storage implementations at once. I've created an issue for it: #45

Comment (HellPat) [inline src/Storage/Session.php:73]: :+1:

**Pull request merged**